# Recognize named entities on Twitter with LSTMs

In this assignment, you will use a recurrent neural network to solve Named Entity Recognition (NER) problem. NER is a common task in natural language processing systems. It serves for extraction such entities from the text as persons, organizations, locations, etc. In this task you will experiment to recognize named entities from Twitter.

For example, we want to extract persons' and organizations' names from the text. Than for the input text:

    Ian Goodfellow works for Google Brain

a NER model needs to provide the following sequence of tags:

    B-PER I-PER    O     O   B-ORG  I-ORG

Where *B-* and *I-* prefixes stand for the beginning and inside of the entity, while *O* stands for out of tag or no tag. Markup with the prefix scheme is called *BIO markup*. This markup is introduced for distinguishing of consequent entities with similar types.

A solution of the task will be based on neural networks, particularly, on Bi-Directional Long Short-Term Memory Networks (Bi-LSTMs).

### Libraries

For this task you will need the following libraries:
 - [Tensorflow](https://www.tensorflow.org) — an open-source software library for Machine Intelligence.
 
In this assignment, we use Tensorflow 1.15.0. You can install it with pip:

    !pip install tensorflow==1.15.0
     
 - [Numpy](http://www.numpy.org) — a package for scientific computing.
 
If you have never worked with Tensorflow, you would probably need to read some tutorials during your work on this assignment, e.g. [this one](https://www.tensorflow.org/tutorials/recurrent) could be a good starting point. 

### Data

The following cell will download all data required for this assignment into the folder `week2/data`.

In [551]:
import sys
sys.path.append("..")
from common.download_utils import download_week2_resources

download_week2_resources()

File data\train.txt is already downloaded.
File data\validation.txt is already downloaded.
File data\test.txt is already downloaded.


### Load the Twitter Named Entity Recognition corpus

We will work with a corpus, which contains tweets with NE tags. Every line of a file contains a pair of a token (word/punctuation symbol) and a tag, separated by a whitespace. Different tweets are separated by an empty line.

The function *read_data* reads a corpus from the *file_path* and returns two lists: one with tokens and one with the corresponding tags. You need to complete this function by adding a code, which will replace a user's nickname to `<USR>` token and any URL to `<URL>` token. You could think that a URL and a nickname are just strings which start with *http://* or *https://* in case of URLs and a *@* symbol for nicknames.

In [552]:
def read_data(file_path):
    tokens = []
    tags = []
    tweet_tokens = []
    tweet_tags = []
    for line in open(file_path, encoding='utf-8'):
        line = line.strip()
        if not line:
            if tweet_tokens:
                tokens.append(tweet_tokens)
                tags.append(tweet_tags)
            tweet_tokens = []
            tweet_tags = []
        else:
            token, tag = line.split()
            
            # Replace all urls with <URL> token
            if token.startswith('http://') or token.startswith('https://'):
                token ="<URL>"
            # Replace all users with <USR> token
            if token.startswith('@'):
                token ="<USR>"
                
            tweet_tokens.append(token)
            tweet_tags.append(tag)
    return tokens, tags

And now we can load three separate parts of the dataset:
 - *train* data for training the model;
 - *validation* data for evaluation and hyperparameters tuning;
 - *test* data for final evaluation of the model.

In [553]:
train_tokens, train_tags = read_data('data/train.txt')
validation_tokens, validation_tags = read_data('data/validation.txt')
test_tokens, test_tags = read_data('data/test.txt')

You should always understand what kind of data you deal with. For this purpose, you can print the data running the following cell:

In [554]:
#print (len(set(train_tokens), len(train_tags))
for i in range(3):
    for token, tag in zip(train_tokens[i], train_tags[i]):
        print('%s\t%s' % (token, tag))
    print()


RT	O
<USR>	O
:	O
Online	O
ticket	O
sales	O
for	O
Ghostland	B-musicartist
Observatory	I-musicartist
extended	O
until	O
6	O
PM	O
EST	O
due	O
to	O
high	O
demand	O
.	O
Get	O
them	O
before	O
they	O
sell	O
out	O
...	O

Apple	B-product
MacBook	I-product
Pro	I-product
A1278	I-product
13.3	I-product
"	I-product
Laptop	I-product
-	I-product
MD101LL/A	I-product
(	O
June	O
,	O
2012	O
)	O
-	O
Full	O
read	O
by	O
eBay	B-company
<URL>	O
<URL>	O

Happy	O
Birthday	O
<USR>	O
!	O
May	O
Allah	B-person
s.w.t	O
bless	O
you	O
with	O
goodness	O
and	O
happiness	O
.	O



### Prepare dictionaries

To train a neural network, we will use two mappings: 
- {token}$\to${token id}: address the row in embeddings matrix for the current token;
- {tag}$\to${tag id}: one-hot ground truth probability distribution vectors for computing the loss at the output of the network.

Now you need to implement the function *build_dict* which will return {token or tag}$\to${index} and vice versa. 

In [555]:
from collections import defaultdict

In [556]:
def build_dict(tokens_or_tags, special_tokens):
    """
        tokens_or_tags: a list of lists of tokens or tags
        special_tokens: some special tokens
    """
    # Create a dictionary with default value 0
    tok2idx = defaultdict(lambda: 0)
    idx2tok = []
    
    # Create mappings from tokens (or tags) to indices and vice versa.
    # At first, add special tokens (or tags) to the dictionaries.
    # The first special token must have index 0.
    
    # Mapping tok2idx should contain each token or tag only once. 
    # To do so, you should:
    # 1. extract unique tokens/tags from the tokens_or_tags variable, which is not
    #    occur in special_tokens (because they could have non-empty intersection)
    # 2. index them (for example, you can add them into the list idx2tok
    # 3. for each token/tag save the index into tok2idx).
    flat_list = [item for sublist in tokens_or_tags for item in sublist]
    distinct_tokens_or_tags = set(flat_list)
    idx2tok = list(set(special_tokens))
    idx2tok.extend(list(distinct_tokens_or_tags - set(special_tokens)))
    tok2idx = {k: v for v, k in enumerate(idx2tok)}
    return tok2idx, idx2tok

After implementing the function *build_dict* you can make dictionaries for tokens and tags. Special tokens in our case will be:
 - `<UNK>` token for out of vocabulary tokens;
 - `<PAD>` token for padding sentence to the same length when we create batches of sentences.

In [557]:
special_tokens = ['<UNK>', '<PAD>']
special_tags = ['O']

# print (train_tokens + validation_tokens)
# Create dictionaries 
token2idx, idx2token = build_dict(train_tokens + validation_tokens, special_tokens)
tag2idx, idx2tag = build_dict(train_tags, special_tags)

The next additional functions will help you to create the mapping between tokens and ids for a sentence. 

In [558]:
print (len(token2idx.keys()))
print(len(tag2idx.keys()))
#print (token2idx['<PAD>'])
def words2idxs(tokens_list):
    return [token2idx[word] for word in tokens_list]

def tags2idxs(tags_list):
    return [tag2idx[tag] for tag in tags_list]

def idxs2words(idxs):
    return [idx2token[idx] for idx in idxs]

def idxs2tags(idxs):
    return [idx2tag[idx] for idx in idxs]

20505
21


### Generate batches

Neural Networks are usually trained with batches. It means that weight updates of the network are based on several sequences at every single time. The tricky part is that all sequences within a batch need to have the same length. So we will pad them with a special `<PAD>` token. It is also a good practice to provide RNN with sequence lengths, so it can skip computations for padding parts. We provide the batching function *batches_generator* readily available for you to save time. 

In [559]:
def batches_generator(batch_size, tokens, tags,
                      shuffle=True, allow_smaller_last_batch=True):
    """Generates padded batches of tokens and tags."""
    
    n_samples = len(tokens)
    if shuffle:
        order = np.random.permutation(n_samples)
    else:
        order = np.arange(n_samples)

    n_batches = n_samples // batch_size
    if allow_smaller_last_batch and n_samples % batch_size:
        n_batches += 1

    for k in range(n_batches):
        batch_start = k * batch_size
        batch_end = min((k + 1) * batch_size, n_samples)
        current_batch_size = batch_end - batch_start
        x_list = []
        y_list = []
        max_len_token = 0
        for idx in order[batch_start: batch_end]:
            x_list.append(words2idxs(tokens[idx]))
            y_list.append(tags2idxs(tags[idx]))
            max_len_token = max(max_len_token, len(tags[idx]))
            
        # Fill in the data into numpy nd-arrays filled with padding indices.
        # print ("INDICES FOR SPECIAL TOKENS:" , token2idx['<PAD>'], tag2idx['O'])
        x = np.ones([current_batch_size, max_len_token], dtype=np.int32) * token2idx['<PAD>']
        y = np.ones([current_batch_size, max_len_token], dtype=np.int32) * tag2idx['O']
        lengths = np.zeros(current_batch_size, dtype=np.int32)
        for n in range(current_batch_size):
            utt_len = len(x_list[n])
            x[n, :utt_len] = x_list[n]
            lengths[n] = utt_len
            y[n, :utt_len] = y_list[n]
        yield x, y, lengths

## Build a recurrent neural network

This is the most important part of the assignment. Here we will specify the network architecture based on TensorFlow building blocks. It's fun and easy as a lego constructor! We will create an LSTM network which will produce probability distribution over tags for each token in a sentence. To take into account both right and left contexts of the token, we will use Bi-Directional LSTM (Bi-LSTM). Dense layer will be used on top to perform tag classification.  

In [560]:
#!python -m pip install tensorflow==2.0
import tensorflow as tf
import numpy as np

In [561]:
# The idea of BiLSTM is to split the state neurons of a regular RNN in a part that is 
# responsible for the positive time direction (forward states) and a part for the negative 
# time direction (backward states). That is take into account both before and after context.
# What came before in sequence and what came after in sequence are considered.

class BiLSTMModel():
    pass

First, we need to create [placeholders](https://www.tensorflow.org/api_docs/python/tf/compat/v1/placeholder) to specify what data we are going to feed into the network during the execution time.  For this task we will need the following placeholders:
 - *input_batch* — sequences of words (the shape equals to [batch_size, sequence_len]);
 - *ground_truth_tags* — sequences of tags (the shape equals to [batch_size, sequence_len]);
 - *lengths* — lengths of not padded sequences (the shape equals to [batch_size]);
 - *dropout_ph* — dropout keep probability; this placeholder has a predefined value 1;
 - *learning_rate_ph* — learning rate; we need this placeholder because we want to change the value during training.

It could be noticed that we use *None* in the shapes in the declaration, which means that data of any size can be feeded. 

You need to complete the function *declare_placeholders*.

In [562]:
def declare_placeholders(self):
    """Specifies placeholders for the model."""

    # Placeholders for input and ground truth output.
    self.input_batch = tf.compat.v1.placeholder(dtype=tf.compat.v1.int32, shape=[None, None], name='input_batch') 
    #self.input_batch = tf.expand_dims(self.input_batch, axis = 2)
    self.ground_truth_tags = tf.compat.v1.placeholder(dtype=tf.compat.v1.int32, shape=[None, None], name='ground_truth_tags') 
  
    # Placeholder for lengths of the sequences.
    self.lengths = tf.compat.v1.placeholder(dtype=tf.compat.v1.int32, shape=[None], name='lengths') 
    
    # Placeholder for a dropout keep probability. If we don't feed
    # a value for this placeholder, it will be equal to 1.0.
    self.dropout_ph = tf.compat.v1.placeholder_with_default(tf.compat.v1.cast(1.0, tf.compat.v1.float32), shape=[])
    
    # Placeholder for a learning rate (tf.compat.v1.float32).
    self.learning_rate_ph = tf.compat.v1.placeholder(dtype= tf.compat.v1.float32, shape=[])

In [563]:
BiLSTMModel.__declare_placeholders = classmethod(declare_placeholders)

Now, let us specify the layers of the neural network. First, we need to perform some preparatory steps: 
 
- Create embeddings matrix with [tf.compat.v1.Variable](https://www.tensorflow.org/api_docs/python/tf/Variable). Specify its name (*embeddings_matrix*), type  (*tf.compat.v1.float32*), and initialize with random values.
- Create forward and backward LSTM cells. TensorFlow provides a number of RNN cells ready for you. We suggest that you use *LSTMCell*, but you can also experiment with other types, e.g. GRU cells. [This](http://colah.github.io/posts/2015-08-Understanding-LSTMs/) blogpost could be interesting if you want to learn more about the differences.
- Wrap your cells with [DropoutWrapper](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper). Dropout is an important regularization technique for neural networks. Specify all keep probabilities using the dropout placeholder that we created before.
 
After that, you can build the computation graph that transforms an input_batch:

- [Look up](https://www.tensorflow.org/api_docs/python/tf/nn/embedding_lookup) embeddings for an *input_batch* in the prepared *embedding_matrix*.
- Pass the embeddings through [Bidirectional Dynamic RNN](https://www.tensorflow.org/api_docs/python/tf/nn/bidirectional_dynamic_rnn) with the specified forward and backward cells. Use the lengths placeholder here to avoid computations for padding tokens inside the RNN.
- Create a dense layer on top. Its output will be used directly in loss function.  
 
Fill in the code below. In case you need to debug something, the easiest way is to check that tensor shapes of each step match the expected ones. 
 

In [564]:
def build_layers(self, vocabulary_size, embedding_dim, n_hidden_rnn, n_tags):
    """Specifies bi-LSTM architecture and computes logits for inputs."""
    
    # Create embedding variable (tf.compat.v1.Variable) with dtype tf.float32
    initial_embedding_matrix = np.random.randn(vocabulary_size, embedding_dim) / np.sqrt(embedding_dim)
    #initial_embedding_matrix = initial_embedding_matrix.astype('float32') 
    embedding_matrix_variable = tf.compat.v1.Variable(initial_value=initial_embedding_matrix, dtype = tf.float32, name="embeddings_matrix")
    # embedding_matrix_variable = tf.compat.v1.Variable(initial_value=tf.compat.v1.random.uniform(shape=[vocabulary_size, embedding_dim]), name="embeddings_matrix", dtype="tf.float32")
    
    # Create RNN cells (for example, tf.compat.v1.nn.rnn_cell.BasicLSTMCell) with n_hidden_rnn number of units 
    # and dropout (tf.compat.v1.nn.rnn_cell.DropoutWrapper), initializing all *_keep_prob with dropout placeholder.
    forward_cell =  tf.compat.v1.nn.rnn_cell.LSTMCell(num_units= n_hidden_rnn)
    backward_cell =  tf.compat.v1.nn.rnn_cell.LSTMCell(num_units= n_hidden_rnn)
    forward_cell = tf.compat.v1.nn.rnn_cell.DropoutWrapper(forward_cell, input_keep_prob=self.dropout_ph,output_keep_prob=self.dropout_ph)
    backward_cell = tf.compat.v1.nn.rnn_cell.DropoutWrapper(backward_cell, input_keep_prob=self.dropout_ph,output_keep_prob=self.dropout_ph)    

    # Look up embeddings for self.input_batch (tf.compat.v1.nn.embedding_lookup).
    # Shape: [batch_size, sequence_len, embedding_dim].
    embeddings = tf.compat.v1.nn.embedding_lookup(embedding_matrix_variable,self.input_batch)
    
    # Pass them through Bidirectional Dynamic RNN (tf.compat.v1.nn.bidirectional_dynamic_rnn).
    # Shape: [batch_size, sequence_len, 2 * n_hidden_rnn]. 
    # Also don't forget to initialize sequence_length as self.lengths and dtype as tf.float32.
    
    # The tf.compat.v1.nn.bidirectional_dynamic_rnn() will return a tuple (outputs, output_state), 
    # where outputs itself is a tuple (output_fw, output_bw), output_fw is a tensor of shape 
    # [batch_size, max_seq_len, cell_fw.output_size] and output_bw is a tensor of shape
    # [batch_size, max_seq_len, cell_bw.output_size]. output_state is also a tuple 
    # (output_state_fw, ouput_state_bw) which gives the forward and backward last hidden state of Bi-LSTM.
    # print (tf.shape(self.input_batch))
    (rnn_output_fw, rnn_output_bw), _ =  tf.compat.v1.nn.bidirectional_dynamic_rnn(cell_fw= forward_cell, cell_bw= backward_cell,inputs= embeddings,  sequence_length=self.lengths, dtype = tf.float32)
    rnn_output = tf.compat.v1.concat([rnn_output_fw, rnn_output_bw], axis=2)

    # Dense layer on top.
    # Shape: [batch_size, sequence_len, n_tags].   
    self.logits = tf.compat.v1.layers.dense(rnn_output, n_tags, activation=None)

In [565]:
BiLSTMModel.__build_layers = classmethod(build_layers)

To compute the actual predictions of the neural network, you need to apply [softmax](https://www.tensorflow.org/api_docs/python/tf/nn/softmax) to the last layer and find the most probable tags with [argmax](https://www.tensorflow.org/api_docs/python/tf/argmax).

In [566]:
def compute_predictions(self):
    """Transforms logits to probabilities and finds the most probable tags."""
    
    # Create softmax (tf.compat.v1.nn.softmax) function
    # default axis = -1; which indicates the last dimension.= n_tags axis
    softmax_output = tf.compat.v1.nn.softmax(self.logits)
    
    # Use argmax (tf.compat.v1.argmax) to get the most probable tags
    # Don't forget to set axis=-1 ,ie, [= n_tags axis]
    # otherwise argmax will be calculated in a wrong way
    self.predictions = tf.compat.v1.argmax(softmax_output, axis =-1)

In [567]:
BiLSTMModel.__compute_predictions = classmethod(compute_predictions)

During training we do not need predictions of the network, but we need a loss function. We will use [cross-entropy loss](http://ml-cheatsheet.readthedocs.io/en/latest/loss_functions.html#cross-entropy), efficiently implemented in TF as 
[cross entropy with logits](https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits_v2). Note that it should be applied to logits of the model (not to softmax probabilities!). Also note,  that we do not want to take into account loss terms coming from `<PAD>` tokens. So we need to mask them out, before computing [mean](https://www.tensorflow.org/api_docs/python/tf/reduce_mean).

In [568]:
def compute_loss(self, n_tags, PAD_index):
    """Computes masked cross-entopy loss with logits."""
    
    # Create cross entropy function function (tf.compat.v1.nn.softmax_cross_entropy_with_logits_v2)
    ground_truth_tags_one_hot = tf.compat.v1.one_hot(self.ground_truth_tags, n_tags)
    # labels should be probability distribution of size (batch_size, sequence_length)
    loss_tensor=tf.compat.v1.nn.softmax_cross_entropy_with_logits_v2(labels= ground_truth_tags_one_hot, logits = self.logits)
    
    mask = tf.compat.v1.cast(tf.compat.v1.not_equal(self.input_batch, PAD_index), tf.compat.v1.float32)
    # Create loss function which doesn't operate with <PAD> tokens (tf.compat.v1.reduce_mean)
    # Be careful that the argument of tf.compat.v1.reduce_mean should be
    # multiplication of mask and loss_tensor.
    self.loss =  tf.compat.v1.reduce_mean(tf.multiply(loss_tensor,mask))

In [569]:
BiLSTMModel.__compute_loss = classmethod(compute_loss)

In [570]:
def perform_optimization(self):
    """Specifies the optimizer and train_op for the model."""
    # print ("learning rate is ",self.learning_rate_ph)
    # Create an optimizer (tf.compat.v1.train.AdamOptimizer)
    self.optimizer =  tf.compat.v1.train.AdamOptimizer(learning_rate = self.learning_rate_ph)
    self.grads_and_vars = self.optimizer.compute_gradients(self.loss)
    
    # Gradient clipping (tf.compat.v1.clip_by_norm) for self.grads_and_vars
    # Pay attention that you need to apply this operation only for gradients 
    # because self.grads_and_vars also contains variables.
    # list comprehension might be useful in this case.
    clip_norm = tf.compat.v1.cast(1.0, tf.compat.v1.float32)
    # Clips tensor values to a maximum L2-norm. - clip_norm value
    self.grads_and_vars1 =  [(tf.compat.v1.clip_by_norm(i[0],clip_norm),i[1]) for i in self.grads_and_vars]
    
    self.train_op = self.optimizer.apply_gradients(self.grads_and_vars1)

The last thing to specify is how we want to optimize the loss. 
We suggest that you use [Adam](https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer) optimizer with a learning rate from the corresponding placeholder. 
You will also need to apply clipping to eliminate exploding gradients. It can be easily done with [clip_by_norm](https://www.tensorflow.org/api_docs/python/tf/clip_by_norm) function. 

In [571]:
BiLSTMModel.__perform_optimization = classmethod(perform_optimization)

Congratulations! You have specified all the parts of your network. You may have noticed, that we didn't deal with any real data yet, so what you have written is just recipes on how the network should function.
Now we will put them to the constructor of our Bi-LSTM class to use it in the next section. 

In [572]:
def init_model(self, vocabulary_size, n_tags, embedding_dim, n_hidden_rnn, PAD_index):
    self.__declare_placeholders()
    self.__build_layers(vocabulary_size, embedding_dim, n_hidden_rnn, n_tags)
    self.__compute_predictions()
    self.__compute_loss(n_tags, PAD_index)
    self.__perform_optimization()

In [573]:
BiLSTMModel.__init__ = classmethod(init_model)

## Train the network and predict tags

[Session.run](https://www.tensorflow.org/api_docs/python/tf/Session#run) is a point which initiates computations in the graph that we have defined. To train the network, we need to compute *self.train_op*, which was declared in *perform_optimization*. To predict tags, we just need to compute *self.predictions*. Anyway, we need to feed actual data through the placeholders that we defined before. 

In [574]:
def train_on_batch(self, session, x_batch, y_batch, lengths, learning_rate, dropout_keep_probability):
    feed_dict = {self.input_batch: x_batch,
                 self.ground_truth_tags: y_batch,
                 self.learning_rate_ph: learning_rate,
                 self.dropout_ph: dropout_keep_probability,
                 self.lengths: lengths}
    
    session.run(self.train_op, feed_dict=feed_dict)
    
    # To check if learning rate is correctly taken in by training
    
    # with tf.compat.v1.Session() as sess1:
        #init = tf.compat.v1.global_variables_initializer()
        #sess1.run(init)
        #print (self.learning_rate_ph.eval(feed_dict=feed_dict))

In [575]:
BiLSTMModel.train_on_batch = classmethod(train_on_batch)

Implement the function *predict_for_batch* by initializing *feed_dict* with input *x_batch* and *lengths* and running the *session* for *self.predictions*.

In [576]:
def predict_for_batch(self, session, x_batch, lengths):
    feed_dict = {self.input_batch: x_batch,
                 self.lengths: lengths}
    # sessions.run(output_target_variable_expected, input_graph_feed)
    predictions = session.run(self.predictions, feed_dict=feed_dict)
    return predictions

In [577]:
BiLSTMModel.predict_for_batch = classmethod(predict_for_batch)

We finished with necessary methods of our BiLSTMModel model and almost ready to start experimenting.

### Evaluation 
To simplify the evaluation process we provide two functions for you:
 - *predict_tags*: uses a model to get predictions and transforms indices to tokens and tags;
 - *eval_conll*: calculates precision, recall and F1 for the results.

In [578]:
from evaluation import precision_recall_f1

In [579]:
def predict_tags(model, session, token_idxs_batch, lengths):
    """Performs predictions and transforms indices to tokens and tags."""
    #Performs predictions
    tag_idxs_batch = model.predict_for_batch(session, token_idxs_batch, lengths)
    
    tags_batch, tokens_batch = [], []
    for tag_idxs, token_idxs in zip(tag_idxs_batch, token_idxs_batch):
        tags, tokens = [], []
        for tag_idx, token_idx in zip(tag_idxs, token_idxs):
            
            # transforms indices or predictions to tokens and tags using the dictionaries
            tags.append(idx2tag[tag_idx])
            tokens.append(idx2token[token_idx])
        tags_batch.append(tags)
        tokens_batch.append(tokens)
    return tags_batch, tokens_batch
    
    
def eval_conll(model, session, tokens, tags, short_report=True):
    """Computes NER quality measures using CONLL shared task script."""
    
    y_true, y_pred = [], []
    for x_batch, y_batch, lengths in batches_generator(1, tokens, tags):
        # tags_batch is the predicted tags for x_batch tokens
        tags_batch, tokens_batch = predict_tags(model, session, x_batch, lengths)
        if len(x_batch[0]) != len(tags_batch[0]):
            raise Exception("Incorrect length of prediction for the input, "
                            "expected length: %i, got: %i" % (len(x_batch[0]), len(tags_batch[0])))
        predicted_tags = []
        ground_truth_tags = []
        for gt_tag_idx, pred_tag, token in zip(y_batch[0], tags_batch[0], tokens_batch[0]): 
            if token != '<PAD>':
                ground_truth_tags.append(idx2tag[gt_tag_idx])
                predicted_tags.append(pred_tag)

        # We extend every prediction and ground truth sequence with 'O' tag
        # to indicate a possible end of entity.
        y_true.extend(ground_truth_tags + ['O'])
        y_pred.extend(predicted_tags + ['O'])
        
    results = precision_recall_f1(y_true, y_pred, print_results=True, short_report=short_report)
    return results

## Run your experiment

Create *BiLSTMModel* model with the following parameters:
 - *vocabulary_size* — number of tokens;
 - *n_tags* — number of tags;
 - *embedding_dim* — dimension of embeddings, recommended value: 200;
 - *n_hidden_rnn* — size of hidden layers for RNN, recommended value: 200;
 - *PAD_index* — an index of the padding token (`<PAD>`).

Set hyperparameters. You might want to start with the following recommended values:
- *batch_size*: 32;
- 4 epochs;
- starting value of *learning_rate*: 0.005
- *learning_rate_decay*: a square root of 2;
- *dropout_keep_probability*: try several values: 0.1, 0.5, 0.9.

However, feel free to conduct more experiments to tune hyperparameters and earn extra points for the assignment.

In [580]:
tf.compat.v1.disable_eager_execution()
tf.compat.v1.reset_default_graph()

# vocab = train +validation unique tokens as in token dict
# tags = train tags as in tags dict
# len(token2idx) = 20505
# len (tag2idx) = 21
# PAD_index = token2idx["<PAD>"]
model = BiLSTMModel(vocabulary_size=20505, n_tags=21, embedding_dim= 200, n_hidden_rnn = 200 ,PAD_index=token2idx["<PAD>"])

If you got an error *"Tensor conversion requested dtype float64 for Tensor with dtype float32"* in this point, check if there are variables without dtype initialised. Set the value of dtype equals to *tf.compat.v1.float32* for such variables.

Finally, we are ready to run the training!

In [581]:
import math

batch_size = 32
n_epochs = 4
learning_rate = 0.01
learning_rate_decay = math.sqrt(2)
dropout_keep_probability = 0.9

sess = tf.compat.v1.Session()
sess.run(tf.compat.v1.global_variables_initializer())

print('Start training... \n')
for epoch in range(n_epochs):
    # For each epoch evaluate the model on train and validation data
    print('-' * 20 + ' Epoch {} '.format(epoch+1) + 'of {} '.format(n_epochs) + '-' * 20)
    print('Train data evaluation:')
    eval_conll(model, sess, train_tokens, train_tags, short_report=True)
    print('Validation data evaluation:')
    eval_conll(model, sess, validation_tokens, validation_tags, short_report=True)
    
    # Train the model
    for x_batch, y_batch, lengths in batches_generator(batch_size, train_tokens, train_tags):
        model.train_on_batch(sess, x_batch, y_batch, lengths, learning_rate, dropout_keep_probability)
        
    # Decaying the learning rate
    learning_rate = learning_rate / learning_rate_decay
    # print ("LR:", learning_rate)
print('...training finished.')

# The number after word "correct" you can find in the log 
# (in both train and test cases) is precision that is the number of correct phrases among found phrases.

Start training... 

-------------------- Epoch 1 of 4 --------------------
Train data evaluation:
processed 105778 tokens with 4489 phrases; found: 78396 phrases; correct: 163.

precision:  0.21%; recall:  3.63%; F1:  0.39

Validation data evaluation:
processed 12836 tokens with 537 phrases; found: 9485 phrases; correct: 17.

precision:  0.18%; recall:  3.17%; F1:  0.34

-------------------- Epoch 2 of 4 --------------------
Train data evaluation:
processed 105778 tokens with 4489 phrases; found: 3728 phrases; correct: 1891.

precision:  50.72%; recall:  42.13%; F1:  46.03

Validation data evaluation:
processed 12836 tokens with 537 phrases; found: 294 phrases; correct: 149.

precision:  50.68%; recall:  27.75%; F1:  35.86

-------------------- Epoch 3 of 4 --------------------
Train data evaluation:
processed 105778 tokens with 4489 phrases; found: 4806 phrases; correct: 3562.

precision:  74.12%; recall:  79.35%; F1:  76.64

Validation data evaluation:
processed 12836 tokens with 537

In [582]:
# The output details of different parameter settings are given below:-
# Overfitting when epoch > 4
# Keep rate = 0.9 (complex so need enough nodes - less dropouts)
# Best is 0.9 drop and 4 epochs.(Based on hypertuning done)

########################################################################################################################

### Parameters:
batch_size = 32
n_epochs = 4
learning_rate = 0.005
learning_rate_decay = math.sqrt(2)
dropout_keep_probability = 0.1

### Output:
Train data evaluation:
processed 105778 tokens with 4489 phrases; found: 2543 phrases; correct: 474.

precision:  18.64%; recall:  10.56%; F1:  13.48

Validation data evaluation:
processed 12836 tokens with 537 phrases; found: 243 phrases; correct: 34.

precision:  13.99%; recall:  6.33%; F1:  8.72
            
########################################################################################################################
########################################################################################################################

### Parameters:
batch_size = 32
n_epochs = 4
learning_rate = 0.005
learning_rate_decay = math.sqrt(2)
dropout_keep_probability = 0.5

### Output:
Train data evaluation:
processed 105778 tokens with 4489 phrases; found: 4543 phrases; correct: 3083.

precision:  67.86%; recall:  68.68%; F1:  68.27

Validation data evaluation:
processed 12836 tokens with 537 phrases; found: 344 phrases; correct: 172.

precision:  50.00%; recall:  32.03%; F1:  39.05
########################################################################################################################
########################################################################################################################

### Parameters:
batch_size = 32
n_epochs = 4
learning_rate = 0.005
learning_rate_decay = math.sqrt(2)
dropout_keep_probability = 0.9

### Output:

Train data evaluation:
processed 105778 tokens with 4489 phrases; found: 4723 phrases; correct: 3731.

precision:  79.00%; recall:  83.11%; F1:  81.00

Validation data evaluation:
processed 12836 tokens with 537 phrases; found: 388 phrases; correct: 172.

precision:  44.33%; recall:  32.03%; F1:  37.19

########################################################################################################################
########################################################################################################################

### Parameters:
batch_size = 32
n_epochs = 5
learning_rate = 0.005
learning_rate_decay = math.sqrt(2)
dropout_keep_probability = 0.9

### Output:

Train data evaluation:
processed 105778 tokens with 4489 phrases; found: 4666 phrases; correct: 4210.

precision:  90.23%; recall:  93.78%; F1:  91.97

Validation data evaluation:
processed 12836 tokens with 537 phrases; found: 406 phrases; correct: 197.

precision:  48.52%; recall:  36.69%; F1:  41.78

########################################################################################################################
########################################################################################################################


Now let us see full quality reports for the final model on train, validation, and test sets. To give you a hint whether you have implemented everything correctly, you might expect F-score about 40% on the validation set.

**The output of the cell below (as well as the output of all the other cells) should be present in the notebook for peer2peer review!**

In [583]:
print('-' * 20 + ' Train set quality: ' + '-' * 20)
train_results = eval_conll(model, sess, train_tokens, train_tags, short_report=False)

print('-' * 20 + ' Validation set quality: ' + '-' * 20)
validation_results = eval_conll(model, sess, validation_tokens, validation_tags, short_report=False)

print('-' * 20 + ' Test set quality: ' + '-' * 20)

vocab = list(token2idx.keys())

# test_tokens, test_tags = read_data('data/test.txt')

for each in test_tokens:
    for j in range(len(each)):
        if each[j] not in vocab:
            each[j]="<UNK>"

test_results = eval_conll(model, sess, test_tokens, test_tags, short_report=False)

# Precision = correct/found - found is predicted phrases by model.
# Recall = correct/phrases in token set. (includes predicted and not predicted)

# Sea has fish, shrimp, etc. Task is to catch only fishes.
# There are x fishes in sea and if less caught but all caught are fishes- high precision = caught fishes/total number of catches
# If many caught & lot many fishes but lot many shrimps etc, recall high = caught fishes/total number of fishes in sea

-------------------- Train set quality: --------------------
processed 105778 tokens with 4489 phrases; found: 4527 phrases; correct: 4413.

precision:  97.48%; recall:  98.31%; F1:  97.89

	     company: precision:   97.40%; recall:   98.91%; F1:   98.15; predicted:   653

	    facility: precision:   95.65%; recall:   98.09%; F1:   96.86; predicted:   322

	     geo-loc: precision:   99.10%; recall:   99.20%; F1:   99.15; predicted:   997

	       movie: precision:   90.28%; recall:   95.59%; F1:   92.86; predicted:    72

	 musicartist: precision:   96.58%; recall:   97.41%; F1:   97.00; predicted:   234

	       other: precision:   96.21%; recall:   97.23%; F1:   96.71; predicted:   765

	      person: precision:   98.87%; recall:   98.98%; F1:   98.93; predicted:   887

	     product: precision:   97.80%; recall:   97.80%; F1:   97.80; predicted:   318

	  sportsteam: precision:   98.15%; recall:   97.70%; F1:   97.92; predicted:   216

	      tvshow: precision:   85.71%; recall:  

### Questions on submission site:

#### Parameters for model
batch_size = 32
n_epochs = 4
learning_rate = 0.01
learning_rate_decay = math.sqrt(2)
dropout_keep_probability = 0.9

Find tags with the best quality on validation and test data.

    >> Validation:company (58.59 F1-score), geo-loc and facility (in F1 score order)
    >> Test: geo-loc (62.50 F1-score) and company (in F1 score order)
    
Find tags with the worst quality on validation and test data.
    
    >> Validation:tvshow, movie
    >> Test:tvshow, movie
    
    
What is the percentage of the correct phrases on test data?

    >> 236/479 found phrases = precision:  49.27%

#### Parameters for model
batch_size = 32
n_epochs = 4
learning_rate = 0.005
learning_rate_decay = math.sqrt(2)
dropout_keep_probability = 0.5

Find tags with the best quality on validation and test data.

    >> Validation:company (58.06 F1-score), geo-loc and facility (in F1 score order)
    >> Test: geo-loc (62.24 F1-score) , company, and facility (in F1 score order)
    
Find tags with the worst quality on validation and test data.
    
    >> Validation:tvshow, movie, sportsteam
    >> Test:tvshow, movie, sportsteam
    
    
What is the percentage of the correct phrases on test data?

    >> 221/436 found phrases = precision is 50.69%
    >> 221/604 total phrases = recall is 36.59%

There could be a variety of reasons that cause the law quality of the model.

A general advice would be:

1. check that you have all indices correct, e.g. that you lookup the right embeddings for each inout batch, etc.
2. carefully read  comments in the notebook and make sure you follow them
3. do debugging as you would normally do in any task, e.g. try printing out some Tensors during the learning to notice any potential badness in there
4. make sure learning rate and decaying looks good.

### Conclusions

Could we say that our model is state of the art and the results are acceptable for the task? Definately, we can say so. Nowadays, Bi-LSTM is one of the state of the art approaches for solving NER problem and it outperforms other classical methods. Despite the fact that we used small training corpora (in comparison with usual sizes of corpora in Deep Learning), our results are quite good. In addition, in this task there are many possible named entities and for some of them we have only several dozens of trainig examples, which is definately small. However, the implemented model outperforms classical CRFs for this task. Even better results could be obtained by some combinations of several types of methods, e.g. see [this](https://arxiv.org/abs/1603.01354) paper if you are interested.